LargePrimeNumbers WIP

Fraudik opened the pull request:
Настроил проект, реализовал вероятностные тесты (больше, скорее всего. сами тесты добавляться не будут). 
Дальше хочу 
- внести изменения при необходимости (пока ещё нет устоявшейся архитектуры);
- в обертках добавить полноценные проверки (с переборами параметров для тех тестов, где они есть и с комбинациями тестов); 
- добавить юнит-тестов (корректность определения простоты на простых числах малого порядка, и одного-парочки достаточно большого; может быть специфических и с замером времени на числах, которые около минуты обрабатываются); 
- либо здесь, либо после следующего этапа -- расширить функционал main (в отдельных файлах разные варианты ввода и  тестирования), главным в этом функционале считаю возможность запускать и замерять время на действительно больших простых числа. Но тут буду советоваться, может лучше отдельные гугл тесты создать и внутри них замерять время
- затем реализовывать детерминированные проверки (вместе с ними добавится что-то для генерации чисел Мерсенна и подобных -- может и тесты дополнятся для вероятностных) и решета (квадратичное в конце, уже как отдельный этап решет).

Fraudik committed:
Add readme, gitignore and clang formats

Fraudik committed:
Add project itself

Fraudik commented:
В contrib субмодуль гугл тестов(

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+set (CMAKE_C_EXTENSIONS ON)
+set (CMAKE_C_STANDARD_REQUIRED ON)
+
+set (DEBUG_INFO_FLAGS "-g -gdwarf-4")
```
Не делай компиляторо зависимые флаги. Либо тогда ставь блок на компилирование всем кроме gcc (или что ты используешь).

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+set (DEBUG_INFO_FLAGS "-g -gdwarf-4")
+
+set (CMAKE_CXX_FLAGS                     "${CMAKE_CXX_FLAGS} ${COMPILER_FLAGS}")
+set (CMAKE_CXX_FLAGS_RELWITHDEBINFO      "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -O3 ${DEBUG_INFO_FLAGS} ${CMAKE_CXX_FLAGS_ADD}")
```
Вот это опять компиляторо зависимые флаги. Например у MSVC его просто нет, но даже -O2 для него будет /O2. Плохая идея так делать. Должна быть возможность либо устанавливать флаги через общий интерфейс cmake либо уже под конкретный компилятор устанавливать флаги если только им собираешь. Ну либо если проект не собирается ничем кроме gcc, то просто блок на другие компиляторы, чтобы проект вообще не мог собраться и сразу падал с сообщением об этом.

Fraudik replied:
Сделал флаги для GNU (на котором собираю), Можно, в целом, и поставить блок на другие (если имеет смысл), но не нашёл как это сделать (искал для cmake запрет определенных компиляторов или разрешение только одного).

DimaTrushin commented on CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+add_subdirectory(programs)
+
+#add_executable(large_prime_numbers main.cpp primality_checks/TrialDivision.cpp primality_checks/TrialDivision.h src/primality_tests_wrapper.h)
+#target_link_libraries(large_prime_numbers gmp gmpxx)
```
Добавляй новые строчки в конце всех файлов. А то на это ругается git и все пользователи репозитория будут видеть от него сообщения при клонировании и использовании.

DimaTrushin commented on README.md:
```diff
@@ -0,0 +1,41 @@
+
+## Структура проекта:
+
+- /contrib -- содержит дополнительные субмодули, на данный момент это googletest
```
У тебя сейчас contrib лежит тупо подпапкой, а не git submodule. Надо перезалить. А то тут от него немеренно кода и файлов. Вот тут глянь как делать https://git-scm.com/book/en/v2/Git-Tools-Submodules

DimaTrushin commented on programs/large_prime_numbers:
А это что за бинарь тут лежит? Это предпосчитанные и сериализованные числа в каком-то формате?

Fraudik replied:
Убрал его, непонятный артефакт сборок. Не знаю, как и от чего появляетсядействие, которое его вызывает, не смог локализовать (билд, изменения CMAKE, просто запуск в разных конфигурациях).

DimaTrushin commented on programs/main.cpp:
```diff
@@ -0,0 +1,37 @@
+  catch (const std::exception &exc)
+  {
+    std::cerr << exc.what();
+  }
```
Если уж ты ловишь exception, то обязательно добавляй в конце catch(...) {} чтобы у тебя неизвестные исключения не утекли в OS.

DimaTrushin commented on tests/probable_primes_test.cpp:
```diff
@@ -0,0 +1,80 @@
+#include <gtest/gtest.h>
+
+#include "primality_tests_wrapper.h"
+#include "probable_primes.cpp"
```
Какогохрена у тебя cpp файлы инклудятся!? Никогда так не делай вообще никогда так не делай. Если тебе надо заинклудить что-то из cpp файла, то делай в нем нужный тебе хедер и подключай хедер.

DimaTrushin commented on programs/main.cpp:
```diff
@@ -0,0 +1,37 @@
+#include "probable_primes.cpp"
+
+// For further development I expect this to move to another file in /programs dir
+void ExampleNumberPrimalityCheck(mpz_t number)
```
1) В cpp файлах никогда не объявляй функции, классы и переменные вне anonimous namespace. Оберни этот метод в него. Только функция main может быть без namespace.
2) И еще это нормально, что mpz_t принимается by value?

Fraudik replied:
1 сделал.
Во 2 это специфика gmp чисел и стремной архитектуры. mpz_t это вообще typedef на struct__(anonymous)*mpz_t (__mpz_struct*). Из себя mpz_struct представляет собой массив с ровно одним элементом (поэтому и 
В итоге тут на самом деле передача by reference

DimaTrushin replied:
Все понял. Тогда хорошо.

DimaTrushin replied:
Тем не менее, саму функцию в cpp файле надо обернуть в anunymous namespace.

DimaTrushin commented on programs/main.cpp:
```diff
@@ -0,0 +1,37 @@
+#include <iostream>
+
+#include "primality_tests_wrapper.h"
+#include "probable_primes.cpp"
```
Не инклудь cpp файлы. Ты обманываешь ожидания пользователей и можешь словить ODR.

DimaTrushin commented on src/primality_tests_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+
+#include <gmpxx.h>
+
+enum PrimalityStatus {
```
1) Не пиши в global namespace. Выбери для проекта namespace и все помести внутрь него.
2) Вместо enum сделай enum class иначе у тебя эти имена уплывают в global namespace. Это очень черевато большими проблемами с линковкой. Я так понимаю тебе эти значения нужны как флаги и ты с ними никаких операций как с масками делать не будешь. Если будешь, то тогда надо сделать enum внутри struct. Тогда имена утекут внутрь struct, но не в окружающий namespace.

Fraudik replied:
Помещать в enum class даже при выборе собственного namespace для проекта? Сейчас сделал так (нужны только как флаги)

DimaTrushin commented on src/primality_tests_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+};
+
+template<class PrimalityTest, class ... TestSpecificArgs>
+bool PrimalityTestWrapper(mpz_t n, PrimalityTest primality_test, TestSpecificArgs...args) {
```
- Если уж писать хороший код на шаблонах, то надо делать через universal reference, потому что вообще говоря и тест и его аргументы могут быть тяжелыми и ты тут принимаешь по значению. Шаблонный аргумент такого вида никогда не ссылка. Как минимум я думаю, что надо написать так
```cpp
template<class PrimalityTest, class ... TestSpecificArgs>
bool PrimalityTestWrapper(mpz_t n, PrimalityTest primality_test, TestSpecificArgs&&... args)  {
...
auto test_result = primality_test(n, std::forward<TestSpecificArgs>(args)...);
```
Это называется perfect forwarding. Этот способ принимает аргументы по ссылке причем по rvalue или lvalue решается автоматически. И в последней строчке происходит правильное кастование ссылок с помощью forward.

- Если тест у тебя тоже может быть тяжелым, то тут стоит и его принять по универсальной ссылке. Тогда надо только сигнатуру изменить.

DimaTrushin commented on src/primality_tests_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+  if (mpz_cmp_ui(n, 2) < 0)
+    return false;
+
+  if (mpz_divisible_ui_p(n, 2)) {
+    if (mpz_cmp_ui(n, 2) == 0)
+      return true;
+    else
+      return false;
+  }
```
Я реально репу стал чесать на тему что тут написано. Почему не так:
```cpp
if (mpz_cmp_ui(n, 2) == 0)
  return true;
if (mpz_divisible_ui_p(n, 2))
  return false;
```
?? Это и короче и понятнее, что ты делаешь.

DimaTrushin commented on src/primality_tests_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+  if (mpz_cmp_ui(n, 2) < 0)
+    return false;
+
+  if (mpz_divisible_ui_p(n, 2)) {
```
А что в mpz такие кривые прототипы на уровне Си функций?

Fraudik replied:
Да, я даже сначала подумал больше про ассемблерные. Зато есть разные приятные функции вроде mpz_powm или mpz_fdiv_q_2exp, которые эффективно выполняют сразу комбинацию действий (да и остальные тоже так оптимизированы, с байтовой магией)

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -0,0 +1,500 @@
+
+#include "primality_tests_wrapper.h"
+
+PrimalityStatus FermatProbablePrimeTest(mpz_t a, mpz_t p) {
```
А у тебя пререквизиты к этим функциям есть? Если есть, может быть их тут указать комментарием перед тестом? Так же стоит поставить assert-ы на эти preconditions. Это может быть разумно из условий производительности, когда тесты вызываются один за другим и мы уже знаем, что preconditions есть и не надо их проверять внутри.

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -0,0 +1,500 @@
+#include <gmpxx.h>
+
+#include "primality_tests_wrapper.h"
```
Вяглядит так, что тебе нужен не этот инклуд, а только инклуд enum-а. Тогда его стоит в отдельный h файл выделить, а то ты тут подтягиваешь лишние вещи. Вроде как эти тесты не должны знать про wrapper, наоборот он как бы должен их видеть и в  себя подставлять.

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -0,0 +1,500 @@
+  mpz_init_set_ui(intermediate_result, 0);
+  mpz_gcd(intermediate_result, a, p);
+
+  // Checking if $gcd(a,p) = 1$
+  if (mpz_cmp_ui(intermediate_result, 1) > 0) {
+    mpz_clear(intermediate_result);
+    return Composite;
+  }
```
Вот это как раз похоже на precondition. Можно его вынести отдельно. И вообще можно написать две версии: проверяющую precondition и не проверяющую его. Первую реализовать через вторую. И у тебя выделится функция для проверки precondition с нормальным именем. То есть ты можешь задать тут вспомогательную функцию с названием are_coprime(m, n) и не надо будет добавлять такой комментарий, название функции тебе сразу выдаст результат.

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -0,0 +1,500 @@
+  }
+}
+
+PrimalityStatus EulerJacobiProbablePrimeTest(mpz_t a, mpz_t p) {
```
Очень странно, что внутри функции ты не боишься вводить длинные названия, а тут однобуквенные.  С одной стороны, так как в математических алгоритмах действительно использовать короткие имена имеет больше смысла для читаемости, то тут хотя бы комментарий надо добавить перед функцией с описанием, что ее аргументы значат. Потому что иначе могут получиться длинные не читаемые формулы. С другой стороны, ты используешь длинные имена для каких-то промежуточных переменных. Причем ты им даешь совершенно неудачные имена вроде intermediate_result.

Вопрос. Это для переиспользования памяти и чтобы не переинициализировать длинку? А ты уверен что там не происходит реаллокации при новых присвоениях? Вообще вот причину почему ты переиспользуешь в разных частях кода одну и ту же переменную стоит пояснять. Дело в том, что это как раз тот элемент дизайна который вызывает вопросы и который стоит объяснить комментарием. Если это вызвано производительностью, то ты измерял какой это дает импакт? На сколько это улучшает производительность?

Fraudik replied:
Однобуквенные оставил так, как они в математических формулах названы, если нет адекватного названия (вроде base, number). У теста Лукаса нет альтернативных осмысленных названий (он просто принимает пару чисел (P, Q) и проверяемое число). 

А как лучше назвать intermediate_result?  Сюда я складывал результаты вычислений по формуле (и ее итога, соответственно), по которой потом делается какой-то вывод (обычно результат проверяется на равенство нулю). Сейчас называю formula_result, если одна формула в тесте, и left_/right_ в начале, если их две
В каких-то местах заменил temp на более явное p_minus_one и не переиспользовал в том же EulerJacobiProbablePrimeTest как потом символ Якоби, в других более комплексных (тест Лукаса) останется таким же скорее всего. 

Да, реалокаций не происходит, mpz функции просто явно перезаписывают первый элемент массивов mpz_struct

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -0,0 +1,500 @@
+PrimalityStatus LucasProbablePrimeTest(mpz_t n, int64_t p, int64_t q) {
+  int64_t d = p * p - 4 * q;
+
+  // Can not produce required Lucas sequence
+  if (d == 0) {
+    return Error;
+  }
```
Может это вынести п пререквизиты метода и поставить тут assert, считая, что программист ошибся? Или ты хочешь именно такое поведение?

Fraudik replied:
С учетом добавления логики ассертов -- переделал. Формула для d в целом простая, можно считать, что программист ошибся.

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -0,0 +1,500 @@
+  return Composite;
+}
+
+PrimalityStatus LucasProbablePrimeTest(mpz_t n, int64_t p, int64_t q) {
```
Вот тут стоит пояснить почему именно такие аргументы. А именно, что они значат и почему типы двух последних int64_t.

Fraudik replied:
По мотивам описания теста от авторов
Given integers P and Q, where P > 0 and  D=P^{2}-4Q, let Uk(P, Q) and Vk(P, Q) be the corresponding Lucas sequences.

То есть последние два это параметры, они предполагаются недостаточно большими, чтобы была необходимость в длинке, и так легче выполнять с ними какие-то операции и отделять от проверяемого числа

DimaTrushin commented on src/primality_tests_wrapper.h:
```diff
@@ -0,0 +1,33 @@
+
+  auto test_result = primality_test(n, args...);
+  if (test_result == Error)
+    throw std::runtime_error("Wrong args for primality test");
```
1) Вот это выглядит как управление логикой программы с помощью exception. Почему ты тут просто не возвращаешь Error наружу? Почему именно кидаешь исключения?
2) Если ты прям заботишься о скорости, то лучше отключить при сборке exception. Дело в том, что на msvc это сильно замедляет run-time за счет второго стека, а для gcc это раздувает бинарь и тоже может отозваться на производительности.

Fraudik replied:
С новой логикой очень красиво убиралась эта часть, но мне не дали отключить исключения из-за их использования в gmp библиотеке
/usr/include/gmpxx.h:1243:53: error: exception handling disabled, use ‘-fexceptions’ to enable
 1243 |       throw std::domain_error ("factorial(negative)");
 
 Если знаешь, как это исправить, будет круто

DimaTrushin commented on tests/probable_primes_test.cpp:
```diff
@@ -0,0 +1,80 @@
+
+namespace {
+
+std::vector<int64_t> low_primes =
```
А почему не сделал файл с этими числами не читаешь при инициализации из файла в локальную переменную. Тогда бы можно было назвать файл primes_2_to_ 7919.txt и читать из него.

И еще большой вопрос почему это данное является глобальным? Если его надо переиспользовать между тестами, то у google tests есть такая возможность. Я только не подскажу конкретно как это делается, надо доки читать. Но я эту фишку видел чуть ли не сразу в описании возможностей.

DimaTrushin reviewed (changes requested):
Добавил несколько общих замечаний по коду.
1) Самое главное убери в git submodule весь сторонний код. Иначе у тебя тут написано не много, а на github все виснет из-за размера папки contrib.
2) Оберни все в namespace.
3) Еще есть общий вопрос с пререквизитам к функциям алгоритмов. Я прокомментировал в одном месте.
4) Есть еще вопрос к переиспользованию переменных. Там точно есть в этом смысл.

И еще замечание. Я если встречаю повторно ошибки, то не комментирую их в надежде, что ты поправишь все по аналогии.

Вообще выглядит неплохо. Но так обычным си прет от кода с mpz =)

Fraudik committed:
Remove directory of googletest

Fraudik committed:
Add submodule with googletest

Fraudik committed:
Corrections after first review

Fraudik committed:
Start epoch of adequate names

Fraudik commented:
Здесь сделал исправления после первого ревью, дальше добавление нового буду делать в отдельной локальной ветке (чтобы не смешивать)
Оставил неразрешенными комментарии, где решение не до конца однозначное, это больше про нейминг и сборку (или просто вопрос)

Fraudik committed:
Add tests, fixes in existing logic

Fraudik committed:
Small adjustments

Add BPSWPrimalityTest, remove usage of typedef for rng

DimaTrushin commented on programs/main.cpp:
```diff
@@ -1,37 +1,37 @@
+        std::cout << "Enter a number to check for primality" << std::endl;
+        exit(1);
+    }
+    mpz_class gmp_number_class(argv[1]);
```
А почему эта штука не под try?

DimaTrushin commented on programs/main.cpp:
```diff
@@ -1,37 +1,37 @@
+    catch (...)
+    {
+        std::cerr << "unknown exception thrown";
+    }
```
По хорошему функция main не должна делать конкретную работу. Она должна делегировать всю работу другим. В том числе и обработку исключений. Тут идет нарушение уровня абстракции, ты на высоком уровне решаешь остальные задачи, а тут у тебя конкретный код. Это решается так.
```cpp
try {
  mpz_class gmp_number_class(argv[1]);
  ExampleNumberPrimalityCheck(gmp_number_class.get_mpz_t());
} catch(...) {
  except::react();
}
```
где react -- функция, которой делегирована обработка исключений. И она определена в отдельной паре h/cpp как-то так:
```cpp
except.h
namespace except {
void react();
}
except.cpp
#include "except.h"

namespace except {
void react() {
  try {
    throw;
  } catch(std::exception& e) {
    std::cerr << exc.what();
  } catch(...) {
     std::cerr << "unknown exception thrown";
  }
}
}
```

DimaTrushin commented on src/CMakeLists.txt:
```diff
@@ -1,3 +1,3 @@
 add_headers_and_sources(primality_tests ${CMAKE_CURRENT_SOURCE_DIR})
-add_library(primality_tests ${primality_tests_headers} ${primality_tests_sources})
+add_library(primality_tests ${primality_tests_headers} ${primality_tests_sources} probable_primes.h primality_status.h)
```
Не надо добавлять header файлы в add_library, только source файлы. Дело в том, что эти файлы используются как единицы трансляции для компиляции, а h файлы не для этого. В этом нет никакого смысла, кроме как лишний код для компилятора для пережовывания.

DimaTrushin commented on src/primality_status.h:
```diff
@@ -0,0 +1,11 @@
+namespace large_prime_numbers
```
Нужны гарды либо #pragma once либо через #ifdef-ы. Иначе у тебя ODR violation при инклуде в две разные единицы трансляции.

DimaTrushin commented on src/primality_tests_wrapper.h:
```diff
@@ -2,32 +2,35 @@
-    else
-      return false;
-  }
+PrimalityStatus BasicPrimalityChecks(mpz_t number)
```
ODR violation. Никогда не пиши функции в h файлы без inline модификатора. И вообще вопрос, зачем эта функция определена в h файле? Почему не вынести в cpp файл?

Fraudik replied:
Она в шаблонной функции ниже используется. Так что могу только их вместе в каком-то tpp объявить (так и сделал + inline)
Импортировать что в .h, что в .tpp из .cpp, вроде как раз не очень.

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -1,500 +1,551 @@
+
+bool are_coprime(mpz_t a, mpz_t b) {
+    mpz_t gcd;
+    mpz_init(gcd);
```
До меня только сейчас дошло, что не так с кодом, который использует mpz_t. Это сишные объекты, которые не имеют поведения ожидаемого в c++. То есть у них нет конструктора, их надо инициализировать. У них нет операторов, а надо использовать intrinsic-и. По-хорошему было бы здорово использовать нормальную обертку вокруг них. Но это такая задумка по поводу общего глобального дизайна.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,103 @@
+namespace large_prime_numbers
+{
+
+class LowPrimesList : public testing::Test
```
Так как ты определяешь класс в cpp файле, то оберни в anonymous namespace.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,103 @@
+        int64_t prime;
+        while (low_primes_file >> prime)
+            low_primes_.push_back(prime);
+        low_primes_file.close();
```
Файловый поток не обязательно закрывать явно, он закроется в десрукторе.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,103 @@
+    {
+        std::ifstream low_primes_file("primes_2_to_99991.txt", std::ifstream::in);
+        assert(low_primes_file);
+        int64_t prime;
```
Мне каждый раз перед последовательными push_back-ами хочется сделать resize единожды для вектора. Для этого правда придется этот размер поместить в первую строчку файла с простыми.

Fraudik replied:
reserve, так понимаю
сделал, но будто какая-то нагрузка на разработчика (если список менять), о которой он должен будет помнить всё время.

DimaTrushin replied:
По-хорошему это надо делать не так. У тебя тут кишки и сопли размазаны, а нужно запаковать все в нормальные объекты.
1) Надо сделать класс FileReader, который будет у тебя открывать поток файловый на чтение (и он в своем конструкторе устанавливает все нужные биты).
2) Перегружаешь оператор operator>>(FileReader&, std::vector<T, alloc<T>>& vec); для чтения вектора. Именно внутри этой имплементации у тебя будет храниться размер и только этот оператор знает, что и как надо сериализовать и десериализовать для вектора. Может быть надо, а может быть не надо.
3) И уже потом везде в другом месте, ты пишешь код вида
```cpp
FileReader file(path_to_file);
if(!file.open())
  return;
std::vector<int64_t> low_primes;
FileReader >> low_primes;
```
И никакой лишней работы у разработчика нет, потому что вся работа делается именно там, где она должна делаться, а не размазываются кишки по всему подряд.
4) По хорошему тебе нужно сделать еще operator>> для LowPrimeList и по хорошему тут в конструкторе ты пишешь
```cpp
FileReader file(path_to_file);
assert(file.open());
file >> *this;
```
Но в целом не очень хорошая идея читать в дефолтном конструкторе из файла. Лучше уж читать там, где этот самый LowPrimeList используется.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,103 @@
+};
+
+template <class PrimalityTest, class... TestSpecificArgs>
+void CheckOnListOfPrimes(const std::string& test_name, const std::vector<int64_t>& list_of_primes,
```
Вот этот шаблон тоже в anonymous namespace надо добавить.

DimaTrushin commented on src/probable_primes.h:
```diff
@@ -0,0 +1,35 @@
+namespace large_prime_numbers
+{
+
+const int64_t kSelfridgeDefaultMaxD = 100'000;
```
ODR violation. Если объявляешь переменную в хедере, то надо писать inline. Либо в случае когда есть constexpr конструктор или встроенный тип, то надо делать constexpr. И еще почему эта константа глобальная в namespace лежит? Такое ощущение, что она у тебя должна быть внутри следующего класса статической constexpr константой.

Fraudik replied:
А у меня нет класса дальше, это функция.  При этом она предполагается и для использования извне, для запуска перебора параметров по Селфриджу. В тестах пример такого.

Сначала еще делал ее как параметр по умолчанию, но тогда при одновременном использовании функции с этим параметром по умолчанию внутри (BPSWPrimalityTest) и извне (тесты) начинаются проблемы, потому что его нельзя объявить сразу и в хедере, и в .cpp (иначе ругается либо функция внутри cpp, либо функция-обертка при вызове извне). Счел оптимальным сделать так, чтобы явно его указывали, а дефолтное значение было явно написано вот так

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -1,500 +1,551 @@
-    if (d >= max_d) {
-      mpz_clear(gmp_d);
-      return LucasTestParameters{p, q, Error};
+    mpz_clear(gmp_d);
```
1) Вот тут у меня вопрос гениальный. А мы обязаны вызывать этот метод? Если да и он не вызывается автоматически. То этот код не устойчив к исключениям и очень хрупкий. С другой стороны ты ловишь исключения в main функции. Это прям противоречие. То есть это означает, что нам нужен raii wrapper для этого типа, иначе мы не можем его нормально использовать.
2) Кроме того, это нарушение уровня абстракции, ты вместо бизнес логики тут занимаешься менеджментом ресурсов.

DimaTrushin commented on src/probable_primes.cpp:
```diff
@@ -1,500 +1,551 @@
+        mpz_add_ui(right_formula_result, right_formula_result, 1);
+    mpz_mod(right_formula_result, right_formula_result, number);
+
+    // Check if $a^((p - 1) / 2) = JacobiSymbol(a, p) mod p$ where $p$ is $number$ and $a$ is $base$
```
Вот такое надо заворачивать в функцию с говорящим названием. Тут надо что-то вроде
```cpp
bool HoldsEulerJacobyCriterion(...);
```
То есть по сути у тебя структура должна быть простая, если мы все завернем в обертки. По типу просто проверили критерий и все.

Fraudik replied:
Немного не понял этот момент. Фактически, здесь вся функция только это и делает: по number и base проверяет критерий. Сначала вычисляется левая сторона, потом правая и они сравниваются. Дробить тут вроде нечего.

Вынес этот комментарий выше, может так это наглядней будет. В остальных тоже так сделал. В целом, пока писал, как раз смотрел, нет ли возможности что-то так вынести, но вроде нет

DimaTrushin reviewed (changes requested):
Смотри Илья. Я увидел главную проблему: класс mpz_t не совместим с c++ парадигмой поведения. Его надо отдельно инициализировать, потом отдельно разрушать. У него нет операторов, а только интринзики. А ты еще и исключения обрабатываешь. Код с mpz_t не безопасен для исключений. Надо руками чистить ресурсы. Это все говорит о том, что его надо завернуть в raii обертку нормальную и уже потом использовать. Это избавит тебя от необходимости вызывать init функции и функции clear для ресурсов. Это будет делать автоматически конструкторы и деструкторы. Так же если ты этим займешься, то будет здорово сделать поддержку арифметических операций с вызовом правильных интринизков. Можно даже запретить явную конвертацию целочисленных типов к нашему классу, чтобы не вызывались дорогие операции. Ну и конечно добавить операторы сравнения. И в итоге у тебя весь код станет высокоуровневым. Тебе не придется в коде с критериями возиться с ресурсами.

В остальном есть только мелкие технические замечания, на которые я указал.

Fraudik committed:
Fixes after second wave

Fraudik commented:
На инклуд random_generator можно пока не обращать внимания, это на будущее (решил не добавлять новый контент в старый ревью).

Fraudik committed:
Adjust style and add some new tests

Add beautiful Enhanced BPSW test from this paper https://homes.cerias.purdue.edu/~ssw/bfw.pdf
Also add some simple determenistic tests

Fraudik committed:
Add subdirs

DimaTrushin commented on programs/except.cpp:
```diff
@@ -0,0 +1,22 @@
+    }
+}
+
+}
```
блять

DimaTrushin commented on programs/main.cpp:
```diff
@@ -0,0 +1,45 @@
+    }
+    try
+    {
+        // here will be call of finite state machine with input checking and starting corresponding test
```
Есть хорошая библиотека для создания state machine
https://github.com/krzysztof-jusiak/sml
Очень рекомендую.

DimaTrushin commented on programs/main.cpp:
```diff
@@ -0,0 +1,45 @@
+    {
+        std::cout << "Enter a number to check for primality" << std::endl;
+        exit(1);
+    }
```
Не пиши мелкую работу в main.

DimaTrushin commented on src/deterministic_tests/CMakeLists.txt:
```diff
@@ -0,0 +1,2 @@
+add_headers_and_sources(deterministic_tests ${CMAKE_CURRENT_SOURCE_DIR})
+add_library(deterministic_tests ${deterministic_tests_sources})
```
Не добавляй хедеры в add_library и add_executable. Они НЕ для этого.

Fraudik replied:
А как тогда? Вот этот момент брал из ШАДовской плюсовой репы. Интересно, в чем тогда лажа

Fraudik replied:
Заменил на add_sources

DimaTrushin commented on src/primality_tests_wrapper.tpp:
Это еще что за расширение? Нахуя, а главное зачем?

Fraudik replied:
Есть практика как использовать .tpp файлы для шаблонов. Разницы в целом нет, так что если ты его не встречал раньше, то убрал

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+class LowPrimesList : public testing::Test
+{
+protected:
+    LowPrimesList()
```
Какого хуя у тебя защищенный конструктор?

Fraudik replied:
Магии Google Test. Хочется использовать такое в тестах, тогда методы должны быть либо protected либо public (тесты по логике гугла наследуются). 
Это из их гайда вроде, могу просто структуру сделать - так  и правда кажется логичней

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+namespace
+{
+
+class LowPrimesList : public testing::Test
```
НЕ используй в названии класса имена стандартных контейнеров. Дело в том, что слово List имеет вполне конкретный смысл в языках программирования. У тебя тут все таки vector, а не list.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+}
+
+template <class PrimalityTest, class... TestSpecificArgs>
+void CheckOnListOfPrimes(
```
Все, что объявлено в cpp файле должно быть в anonymous namespace!

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+
+template <class PrimalityTest, class... TestSpecificArgs>
+void CheckOnListOfPrimes(
+    const std::string& test_name, const std::vector<int64_t>& list_of_primes, PrimalityTest primality_test, TestSpecificArgs&&... args)
```
А нахуя тебе класс LowPrimeList, если ты тут напрямую вектор передаешь?

Fraudik replied:
Чтобы передавать сюда любой вектор простых чисел?
Ну вот у меня есть тесты просто с небольшими простыми числами. Вот тесты с простыми числами Мерсенна. Может еще какие.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+    const std::string& test_name, const std::vector<int64_t>& list_of_primes, PrimalityTest primality_test, TestSpecificArgs&&... args)
+{
+    bool check_result;
+    for (mpz_class checked_prime : list_of_primes)
```
А нахуя ты тут бежишь по элементам НЕ того типа? у тебя же int64_t лежит в контейнере.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+void CheckOnListOfPrimes(
+    const std::string& test_name, const std::vector<int64_t>& list_of_primes, PrimalityTest primality_test, TestSpecificArgs&&... args)
+{
+    bool check_result;
```
не заводи переменную так рано, спрячь в тело цикла.

DimaTrushin commented on tests/low_primes_probable_primes_test.cpp:
```diff
@@ -0,0 +1,129 @@
+    bool check_result;
+    for (mpz_class checked_prime : list_of_primes)
+    {
+        check_result = PrimalityTestWrapper(checked_prime, primality_test, std::forward<TestSpecificArgs>(args)...);
```
Что вообще хранит эта переменная check_result? Что ее имя значит? Заменить на нормальное имя. Почему она bool, а не enum class со статусом.

DimaTrushin commented on src/utilities.h:
```diff
@@ -0,0 +1,22 @@
+
+TwoPowersDecomposition decompose_in_powers_of_two(const mpz_class& a);
+
+inline void mod(mpz_class& number, const mpz_class& mod_base) {
```
Зачем эта функция размещена в хедере?

Fraudik replied:
Потому что она инлайн?
Note: It’s imperative that the function’s definition (the part between the {...}) be placed in a header file, unless the function is used only in a single .cpp file. In particular, if you put the inline function’s definition into a .cpp file and you call it from some other .cpp file, you’ll get an “unresolved external” error from the linker.

Fraudik replied:
Это маленькая функция, которую я часто вызываю в разных местах. Фактически, просто сокращение, чтобы не звать постоянно вот ту конструкцию внутри

DimaTrushin commented on src/utilities.cpp:
```diff
@@ -0,0 +1,24 @@
+
+TwoPowersDecomposition decompose_in_powers_of_two(const mpz_class& a)
+{
+    TwoPowersDecomposition decomposition{};
```
убрать {}, зачем они тут?

DimaTrushin commented on src/utilities.cpp:
```diff
@@ -0,0 +1,24 @@
+    TwoPowersDecomposition decomposition{};
+    decomposition.power_of_two = mpz_scan1(a.get_mpz_t(), 0);
+    // $odd_multiplier = floor(a / 2^power_of_two)$
+    mpz_fdiv_q_2exp((decomposition.odd_multiplier).get_mpz_t(), a.get_mpz_t(), decomposition.power_of_two);
```
Вообще все выглядит так, что тебе mpz_class надо было обернуть в нормальный класс с перегруженными операциями.

Fraudik replied:
Это даже обсудили с тобой, потому что и правда такой момент. Вообще в mpz_class есть перегрузки всего стандартного. Не обернуты только комплексные штуки, которые очень вкусно реализованы внутри (и хочется использовать их, а не писать свои такие же). Вот это уже оборачивать только как такие же функции. Что-то вида
`decomposition.odd_multiplier.fdiv_q_2exp(a, decomposition.power_of_two);`
выглядит тоже как-то странно.

DimaTrushin commented on src/utilities.cpp:
```diff
@@ -0,0 +1,24 @@
+{
+    mpz_class gcd;
+    mpz_gcd(gcd.get_mpz_t(), a.get_mpz_t(), b.get_mpz_t());
+    if (gcd > 1)
+        return false;
+    return true;
```
это называется
```cpp
return !(gcd > 1);
```

Fraudik replied:
Шит.
Это мой злой брат-близнец.

DimaTrushin commented on src/random_generator.h:
```diff
@@ -0,0 +1,33 @@
+
+    // generate value in range [low_threshold; high_threshold]
+    mpz_class generateValue() {
+        return generator_engine_.get_z_range(low_threshold_ + high_threshold_ + 1) - low_threshold_;
```
А почему НЕ + low_threshold_?

DimaTrushin commented on src/random_generator.h:
```diff
@@ -0,0 +1,33 @@
+        generator_engine_.seed(seed);
+    }
+
+    // generate value in range [low_threshold; high_threshold]
```
равновероятно?

Fraudik replied:
Практически, там Вихрь Мерсенна.  Генерирует равномерные последовательности

DimaTrushin commented on src/random_generator.h:
```diff
@@ -0,0 +1,33 @@
+
+#include <gmpxx.h>
+#include <random>
+#include <optional>
```
убрать лишние инклуды.

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.h:
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <gmpxx.h>
+#include <cassert>
+
+#include "../primality_status.h"
+#include "../random_generator.h"
+#include "../utilities.h"
+#include "lucas_pseudoprimes.h"
```
Зачем тут все эти инклуды? Тут тебе достаточно заинклудить только хедер с объявлением mpz_class. Все остальное перенести в cpp файл.

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+PrimalityStatus FermatProbablePrimeTest(const mpz_class& number, const mpz_class& base)
+{
+    mpz_class formula_result;
+    mpz_class exponent{number};
```
А зачем тут копировать number? И что мешало написать
```cpp
mpz_class exponent = number - 1;
```
???

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+    mpz_class formula_result;
+    mpz_class exponent{number};
+    exponent -= 1;
+    mpz_powm(formula_result.get_mpz_t(), base.get_mpz_t(), exponent.get_mpz_t(), number.get_mpz_t());
```
Почему нельзя было обернуть mpz_class в нормальный класс типо class LongInt и обернуть все эти ужасные вызовы функций? Тогда тут ты бы писал нормальный код.

Fraudik replied:
Как выше

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+
+PrimalityStatus FermatProbablePrimeTestWithCheck(const mpz_class& number, const mpz_class& base)
+{
+    assert(are_coprime(number, base));
```
Это НЕ проверка на взимную простоту. assert нужен НЕ для выполнения содержательных задач. Он служит для проверки инвариантов, чтобы отлавливать свои ошибки. Если ты хочешь тест ферма с проверкой, то ты ставишь if, если числа не взаимно просты, то значит возвращаешь статус не простое, и если взаимно просты, то вызываешь тест без проверки. Ты тут все в кашу намешал.

Fraudik replied:
Это инвариант, если программист подал число и базу как не взаимно простые числа, то он дурак. Пререквизит теста
Уточнил про это в чате с контекстом, сказал, что так можно, да

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+    mpz_divexact_ui(exponent.get_mpz_t(), exponent.get_mpz_t(), 2);
+    mpz_powm(left_formula_result.get_mpz_t(), base.get_mpz_t(), exponent.get_mpz_t(), number.get_mpz_t());
+
+    mpz_class right_formula_result{number};
+    int jacobi_symbol = mpz_jacobi(base.get_mpz_t(), number.get_mpz_t());
+    if (jacobi_symbol == -1)
+        right_formula_result -= 1;
+    else if (jacobi_symbol == 1)
+        right_formula_result += 1;
+    mod(right_formula_result, number);
```
У меня иногда такое чувство, что люди разучились писать код на том же уровне, на котором потерян секрет булатной стали. Что мешало написать это понятнее и короче?
```cpp
int jacobi_symbol = mpz_jacobi(base.get_mpz_t(), number.get_mpz_t());
mpz_class right_formula_result = (jacobi_symbol == 1 ? 1 : number - 1);
```
Ты и так используешь кучу if-ов, так зачем вызывать долгий и медленный mod, если ты всего лишь хочешь написать 1 или p-1???

Fraudik replied:
А, про mod правда.
Про тернарные операторы -- увы, на такую конструкцию ругаются 
`Incompatible operand types ('int' and '__gmp_expr<__mpz_struct[1], __gmp_binary_expr<__gmp_expr<__mpz_struct[1], __mpz_struct[1]>, long, __gmp_binary_minus>>')`

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+    mpz_powm(formula_result.get_mpz_t(), base.get_mpz_t(), s.get_mpz_t(), number.get_mpz_t());
+    if (formula_result == 1 || formula_result == number_minus_one)
+        return PrimalityStatus::ProbablePrime;
+    // Check if $a^(2^t * s) mod n = -1 (n - 1)$ for every integer t in (0; r) where $n$ is $number$ and $a$ is $base$
```
запиши по нормальному равенство по модулю, а то хер пойми, что ты тут проверяешь.

DimaTrushin commented on src/probable_prime_tests/probable_primality_tests.cpp:
```diff
@@ -0,0 +1,101 @@
+{
+    mpz_class number_minus_one{number - 1};
+    auto decomposition = decompose_in_powers_of_two(number_minus_one);
+    uint64_t r = decomposition.power_of_two;
```
у тебя везде идут супер длинные названия переменных, а тут внезапно r. Какого хуя?

Fraudik replied:
А s не смутил? Или формула выше, по которой вычисления?
a^(2^t * s) mod n = -1 аor every integer t in [0; r)
Вот как мне здесь назвать t и s и зачем?

Могу r назвать как max_power_of_two. Но тогда чуть сложнее при чтении сверять с формулой (и на 62 строчке как раз показывается, что это логически - как и на 63 про t)

Fraudik replied:
Названия осмысленные и длинные там, где они не упомянуты в формулах и нужно показать -- а что это такое внезапное - а, это промежуточный результат вычислений

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.h:
```diff
@@ -0,0 +1,43 @@
+    PrimalityStatus pre_test_status;
+};
+
+constexpr int64_t kSelfridgeDefaultMaxD = 100'000;
```
добавь inline

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.h:
```diff
@@ -0,0 +1,43 @@
+    mpz_class V_n;
+};
+
+SelfridgeDetail::LucasTestParameters CalculateSelfridgeParametersForLucasTest(const mpz_class& number, int64_t max_d);
```
Раз у тебя структура LucasTestParameters используется как возвращаемое значение, то плохо ее прятать внутри namespace detail. Потому что возвращаемый тип это часть публичного интерфейса твоей функции и пользователь твоей функции будет каждый раз ебаться с тем, что ему оказывается надо через какое-то служебное пространство имен достучаться до этой структуры.

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    mpz_class gmp_q(q);
+    assert(are_coprime(number, gmp_q));
+
+    mpz_class gmp_d = p * p - 4 * q;
```
что если правая часть переполнится?

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    assert(are_coprime(number, gmp_d));
+
+    // bool type for putting function into assert
+    return true;
```
Что это за хуйня? У тебя функция возвращает тождественный true! Что тут происходит? Какого хуя?

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    else if (ret == 1)
+        delta_n -= 1;
+
+    // Calculating Lucas sequences
```
Если ты даешь имя блоку кода, то лучше это выделить в функцию calculateLucasSequences() и это поймет не только программист, но и компилятор.

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    int64_t d = p * p - 4 * q;
+    mpz_class gmp_d{d};
+
+    // $delta_n = n - JacobiSymbol(D/n)$
+    mpz_class delta_n{number};
+    int ret = mpz_jacobi(gmp_d.get_mpz_t(), number.get_mpz_t());
+    if (ret == -1)
+        delta_n += 1;
+    else if (ret == 1)
+        delta_n -= 1;
```
Какого хуя? Ты в комментарии написал, что delta_n = n - jacobi. Так что мешало тебе тут написать
```cpp
int jacobi = mpz_jacobi(gmp_d.get_mpz_t(), number.get_mpz_t());
mpz_class delta_n = n - jacobi;
```
???

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    int64_t d = p * p - 4 * q;
+    mpz_class gmp_d{d};
+
+    // $delta_n = n - JacobiSymbol(D/n)$
```
Выглядит как повторение кода начиная со строчки 44.

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    else if (ret == 1)
+        delta_n -= 1;
+
+    // Calculating Lucas sequences
```
Выгядит как повторение кода со строчки 52. Ты специально размножаешь  код чем больше, тем лучше?

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    auto decomposition = decompose_in_powers_of_two(delta_n);
+    uint64_t r = decomposition.power_of_two;
+    mpz_class t = decomposition.odd_multiplier;
+    // Further will be checked if $U_t = 0 (mod n)$ or exists r in [0; s) that $V_(t * 2^r) = 0 (mod n)$
```
Кто такое U_t??? У тебя нигде нет этой переменной тут. У тебя комментарии врут максимально про все, что только можно.

Fraudik replied:
А как тут назвать, если U_t это член последовательности U с индексом t, а u_h, u_l -- участвующие в вычислении переменные (по формулам и из алгоритмов, рекурсивное вычисление)

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+
+    mpz_class u_h{1}, v_l{2}, v_h{p}, q_l{1}, q_h{1};
+    size_t j;
+    for (j = mpz_sizeinbase(t.get_mpz_t(), 2) - 1; j >= 1; --j)
```
Нарушиение уровня абстракции. Что за хуйню ты тут считаешь?

Fraudik replied:
Внезапно не понял. Это то же, что было выше (исправил дублирование) -- вычисление последовательностей Лукаса.

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    mpz_class gmp_d{d};
+
+    int jacobi_symbol;
+    while (true)
```
Блять. Еще раз увижу while(true)...

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    int64_t d = 5, p = 1, q;
+    mpz_class gmp_d{d};
+
+    int jacobi_symbol;
```
Нахуя ты заводишь переменную за 10 километров от места ее использования??? Вводи переменную там, где ты ее используешь, в данном случае внутри цикла.

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+        }
+
+        // If not found such d in range to max_d
+        if (d >= max_d)
+            assert(d >= max_d);
```
Это что за хуйня??? Ты вообще понимаешь, что этот код делает?

Fraudik replied:
Нет, мне поплохело

DimaTrushin commented on src/probable_prime_tests/lucas_pseudoprimes.cpp:
```diff
@@ -0,0 +1,507 @@
+    }
+
+    q = (1 - d) / 4;
+    // Method A*
```
Что этот комментарий тут значит?

DimaTrushin commented on src/primality_tests_wrapper.tpp:
```diff
@@ -0,0 +1,61 @@
+inline PrimalityStatus BasicPrimalityChecks(const mpz_class& number)
+{
+    if (number < 2)
+        return PrimalityStatus::Composite;
```
Это не верный статус. Во-первых, у тебя число может быть отрицательным. Во вторых, числа -1, 0, 1 не относятся к простым или не просстым. Числа 1 и -1 обратимы, а простым является число, которое не ноль и не обратимо, как и составное.

Fraudik replied:
Поправил, для отрицательных поставил assert-ы. Все же от библиотеки хочется работы с простыми числами (отрицательные можно просто по модулю взять - и без этого тесты с ними работать не будут). Хочется assert и на > 1 поставить, но это уже будто менее очевидно

DimaTrushin commented on src/primality_tests_wrapper.tpp:
```diff
@@ -0,0 +1,61 @@
+namespace large_prime_numbers
+{
+
+inline PrimalityStatus BasicPrimalityChecks(const mpz_class& number)
```
А почему это называется basic? Это названиек никак не отражает сути метода, ты по сути проверяешь на четность и на извлечение корня.

Fraudik replied:
Имел в виду что-то простое, что делается первым делом при проверках на простоту чисел. К тому же тесты как раз почти всегда требуют "нечетные числа", часть такие же требования предъявляют к извлечению корня. В итоге получаются базовые проверки, которым числа должны удовлетворять до дальнейшего тестирования

Fraudik replied:
Подумал про Common..., но получается хуже. Не понял, что смутило - когда я думаю, как объяснить, что делает метод, я прямо говорю "базовые проверки, которые нужно осуществлять перед любой другой проверкой"

Fraudik replied:
Основные проверки - тоже звучит несколько странно. Могу назвать PrimalityChecksBeforeTests

Но и это не очень, так как в проверка буквально тесты простоты. По той же логике, по какой у нас тест Ферма так называется

DimaTrushin commented on src/primality_tests_wrapper.tpp:
```diff
@@ -0,0 +1,61 @@
+}
+
+template <class PrimalityTest, class... TestSpecificArgs>
+bool PrimalityTestWrapper(const mpz_class& number, PrimalityTest primality_test, TestSpecificArgs &&... args)
```
Я уже не в первый раз вижу подобный wrapper. Почему их так много? Они повторяют друг друга?

Fraudik replied:
Первый про то, что мы четко заданные параметры даем -- это часто бывает нужно, особенно для каких-то устоявшихся тестов вроде с базой 2.
Второй же позволяет просто дать число и оставить генерацию базы на генератор случайных чисел, для чего его. собственно, и писал.  Отличаются обертки достаточно сильно - и по логике, и по сигнатурам. Там общего только вызов теста и базовых проверок в начале

DimaTrushin commented on src/primality_tests_wrapper.tpp:
```diff
@@ -0,0 +1,61 @@
+    PrimalityStatus test_result = BasicPrimalityChecks(number);
+    if (test_result == PrimalityStatus::ProbablePrime)
+        test_result = primality_test(number, std::forward<TestSpecificArgs>(args)...);
+    if (test_result == PrimalityStatus::Composite)
+        return false;
+    return true;
```
Блять.

DimaTrushin commented on src/deterministic_tests/deterministic_primality_tests.h:
```diff
@@ -0,0 +1,26 @@
+// сверить сканы на биты (scan1 и scan0) и побитовые штуки
+// разедлить на файлы
+
+PrimalityStatus LucasLehmerPrimalityTest(const mpz_class& mersenne_number, const mp_bitcnt_t& mersenne_power);
```
Почему ты тут пользуешься какими-то сырыми типами вроде mp_bitcnt_t??

Fraudik replied:
Вообще это то же, что и size_t в плане логики наименования и работы  xD
 Currently this is always an unsigned long, but on some systems it will be an unsigned long long in the future. 

 Но тут воспользуюсь принципом: не понятно сразу при первом взгляде на тип -- выкидываем нафиг. Заменил на привычный size_t

DimaTrushin commented on src/deterministic_tests/deterministic_primality_tests.h:
```diff
@@ -0,0 +1,26 @@
+
+#include "../primality_status.h"
+#include "../utilities.h"
+
```
Убрать лишние хедеры в cpp файл.

DimaTrushin reviewed (changes requested):
Прошелся еще раз по этому чуду.

Fraudik committed:
Add MPQS and some fixes

Fraudik commented:
Сделал несколько исправлений по замечаниями (не все, еще буду доделывать) + написал MPQS по той основной книге-источнику, которую ты скидывал.
Жесть, я почему-то думал, что его будет проще и быстрее написать, а больше всего с матрицами нужно будет делать. Зря откладывал на конец - плюс несколько неудобно было по той книге писать. Тем не менее, оно вроде как готово и работает (парочку пробных тестов написал и проверил). Надеюсь, даже не слишком много ляпов сделал - старался писать, учитывая все предыдущие комментарии (и ошибки)

Fraudik commented:
Дальше исправлю остальные замечания и отправлю на ревью. И дальше буду тесты и бенчмарки писать

Fraudik committed:
Style fixes with optional

Making "optional" aspect more explicit

Fraudik committed:
Fixes after review

Main: change add_headers_and_sources to add_sources

Fraudik committed:
Propose fixes after review

Fraudik committed:
Mode includes

Fraudik committed:
Remove programs folder

Decisive move

Fraudik committed:
Remove unused

Fraudik commented:
Итак, я понял, что если пишется библиотека, то мне не нужно писать машину состояний (благо и не начинал). А по времени на это тоже уйдет время и лучше поступить не так, как с квадратичным решетом, и потратить больше времени на отчет и тесты с бенчмарками. А также улучшение кода.

План дальнейших действий:
1) Составить бенчмарки (написав нужный для этого код)
2) Написать отчет
3) Сделать еще исправления, если нужны будут
4) Улучшить покрытие тестами

Причем какую-то часть 4 пункта я бы хотел взять на май - это как раз то, что я могу уже сам сделать без дальнейшего ревью (больше там никаких архитектурных решений и вообще чего-либо странного даже в теории не будет). Ну и это самая рутинная часть работы с самим кодом (но хочется максимального покрытия).

Fraudik committed:
Add FileReader

Fraudik committed:
Rename and add benchmark submodule

Fraudik committed:
Fix naming and replace vector<bool>

Missed performance issues of vector<bool>
https://en.cppreference.com/w/cpp/container/vector_bool
https://isocpp.org/blog/2012/11/on-vectorbool
Got x2 boost on particular test (from ~200s to ~100s on 5-10 runs each)

Fraudik committed:
Add benchmarks

Fraudik committed:
Update utility.h

Fraudik commented:
Все юнит тесты собираюсь добавить сегодня (03.06.2024).
Комиссия будет 06.06

Fraudik commented:
По поводу utilities.h:
```
А почему может такое быть, что вот есть файлы utilities.cpp и utilities.h
И оттуда в другой файл в своей папке импортируется функция. Если функция объявлена в .h файле и определена в .cpp, то выдает ошибку "undefined reference". То есть будто с сигнатурой что-то не так.
Но вот если перенести функцию в .h файл с inline квалификатором, то ошибка исчезает.
Причем в других файлах такой проблемы с инклудом этой же функции не возникает. Все папки абсолютно одинаково устроены с CMakeLists по одной схеме. На верхнем уровне в добавлении именно этой папки тоже ничего не отличается от добавления других
````
Ошибку выдает линковщик, кеш чистый. Ты говорил, что посмотришь в репозитории, что может не так быть

**UPD**: 
ничего не понял, но бага почему-то исчезла. На всякий случай сбросил все кеши (и Cmake), перезагрузил проект и пересобрал все -- заново не воспроизвелась. 
Через какое-то время опять начала появляться, хотя те файлы не трогал. Собираю через утилиту make в WSL. 
inline работают, не inline -- ругается.

Fraudik committed:
Fix mistype

Fraudik committed:
Update tests and filenames, optimize imports

Fraudik committed:
Refactor benchmarks names

Fraudik committed:
Update README.md
